### PR TITLE
fix(net-stubbing): fix waiting on responses with no-body status codes

### LINKED ISSUE #8934: cy.route2 times out waiting for image load
Cypress v5.4.0 and even https://github.com/cypress-io/cypress/pull/8916

Test in https://github.com/cypress-io/cypress-example-recipes/blob/master/examples/stubbing-spying__route2/cypress/integration/image-spec.js#L4

```js
  it.only('spies on loading a static image', () => {
    cy.route2('/images').as('image')
    cy.visit('/pics.html')
    // how to check if the /image route was called once?
    cy.wait('@image')
  })
```

The image request happens, but the test fails it. Notice that the routes tab shows 1 matching request

wait-on-image-times-out

### LINKED ISSUE #8999: Route returning 204 (No content) can't be waited for with route2
### Current behavior

I'm trying to wait for this route:

```ts
 cy.route2({
  method: "DELETE"
}).as("delete");
```

My server returns a response with `204` status (a valid code for delete cf.https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE). I can see the route is intercepted thanks to the _route_ section in the runner but `wait` for this route alias always times out.

If I change the server to return a `200` (even without a body) then the `wait` command passes. Interestingly, in the runner's panel, the request still shows a progress bar even after the `wait` passes:

with a 204 status:

with a 200:

I found a workaround by adding a route handler like this:

```ts
(req) => {
  req.reply(() => {});
  }
```

### Desired behavior

`204` should get the same behaviour as `200`

### Test code to reproduce

```js
 cy.route2({
  method: "DELETE"
}).as("delete");

// trigger the delete somehow

cy.wait("@delete");
```

### Versions

Cypress 5.5.0

### PR DESCRIPTION
**Pull request by flotwig**
- Closes #8999
- Closes #8934 

### User facing changelog

- Fixed an issue where HTTP responses that cannot have a body (like HTTP 304 and HTTP 204) could not be awaited using `cy.intercept`.
- Fixed an issue where HTTP redirects could not be awaited using `cy.intercept` unless dynamically intercepted.

### Additional details

- tricky because internally, Node.js treats 304s differently from 204s
- also tricky because the fix for #4358 had to be maintained or else the 5 second timeout continues

### PR Tasks

- [x] test from RWA issue has been resolved https://github.com/cypress-io/cypress/pull/9097#issuecomment-736845181
- [x] Have tests been added/updated?
- [x] Has the original issue or this PR been tagged with a release in ZenHub? 
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [na] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Comment by eilgin** [comment]
Hello @flotwig, i've just seen that cypress is now launching [6.0](https://docs.cypress.io/guides/references/changelog.html#6-0-0) (_congrats!_). I'd like to know if this PR is still relevant?

**Comment by flotwig** [comment]
> Hello @flotwig, i've just seen that cypress is now launching [6.0](https://docs.cypress.io/guides/references/changelog.html#6-0-0) (_congrats!_). I'd like to know if this PR is still relevant?

@eilgin yep, this is still a work in progress, i hope to get it in the next release

File: packages/network/lib/http-utils.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import _ from 'lodash'
+import { IncomingMessage } from 'http'
+
+// https://github.com/cypress-io/cypress/issues/4298
+// https://tools.ietf.org/html/rfc7230#section-3.3.3
+// HEAD, 1xx, 204, and 304 responses should never contain anything after headers
+const NO_BODY_STATUS_CODES = [204, 304]
+
+export function responseMustHaveEmptyBody (req: IncomingMessage, res: IncomingMessage) {
+  return _.some([_.includes(NO_BODY_STATUS_CODES, res.statusCode), _.invoke(req.method, 'toLowerCase') === 'head'])
```
Comment (bahmutov) [inline packages/network/lib/http-utils.ts:10]: wow, _some_ people write such interesting code just to avoid `||`

Comment (flotwig) [reply packages/network/lib/http-utils.ts:10]: job security

Comment (flotwig) [inline packages/network/lib/http-utils.ts:10]:
```suggestion
  return _.includes(NO_BODY_STATUS_CODES, res.statusCode) || _.invoke(req.method, 'toLowerCase') === 'head']
```

**Comment by kevinold** [comment]
@flotwig It appears that 302 POST's might also be grouped into this.  See this failure when attempting to `cy.intercept` for `/logout` in the RWA.

https://github.com/cypress-io/cypress-realworld-app/pull/676/commits/1b54220efaa4e470dc80f3f327d9e53c08d31ee2

Test to reproduce the issue: https://github.com/cypress-io/cypress-realworld-app/pull/676/files#diff-8781249f032a90ab825382873828906eed104c18c2a93d20e9de10b4872b0df6R143

Screen Shot 2020-12-01 at 3 44 37 PM

**Review by bahmutov — APPROVED** [review]
Works, verified against 304 in https://github.com/cypress-io/cypress-example-recipes/pull/598

**Comment by flotwig** [comment]
> @flotwig It appears that 302 POST's might also be grouped into this. See this failure when attempting to `cy.intercept` for `/logout` in the RWA.

nice, i was able to reproduce this with a simple `cy.intercept('/redirect')`, and it led me to fixing the underlying logic issue that was causing #8999. This PR should also fix waiting on redirects.

Funnily enough, if you intercept the redirect, it does work:

https://github.com/cypress-io/cypress/blob/c0e14ea037e768c8987b4f78000baf51f75fd496/packages/driver/cypress/integration/commands/net_stubbing_spec.ts#L1274-L1290

**Review by bahmutov — APPROVED** [review]
nice

**Pull request merged**